SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -3531,6 +3533,133 @@ bool USpatialNetDriver::HasTimedOut(const float Interval, uint64& TimeStamp)
+{
+	TArray<FNetworkObjectInfo*> TmpConsiderList;
+	TmpConsiderList.Reserve(GetNetworkObjectList().GetActiveObjects().Num());
+
+	Super::ServerReplicateActors_BuildConsiderList(TmpConsiderList, ServerTickTime);
+
+	// Cache if we are running with Spatial Networking to avoid getting the settings in every iteration of the loop.
+	const bool bSpatialNetworking = GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking();
+	const bool bEnableSpatialDataDebugger = GetDefault<USpatialGDKSettings>()->IsSpatialDataDebuggerEnabled();
+
+	for (FNetworkObjectInfo* const& ActorInfo : TmpConsiderList)
+	{
+		AActor* Actor = ActorInfo->Actor;
+		check(Actor != nullptr);
+
+		Worker_EntityId EntityId = PackageMap->GetEntityIdFromObject(Actor);
+
+		// Check for unauthorised data changes no non-auth actors and add auth actors only to the consider list
+		if (bSpatialNetworking && Actor->HasAuthority())
+		{
+			OutConsiderList.Add(ActorInfo);
+		}
+		else if (bSpatialNetworking && bEnableSpatialDataDebugger)
+		{
+			CheckUnauthorisedDataChanges(EntityId, Actor);
+		}
+	}
+}
+
+void USpatialNetDriver::CheckUnauthorisedDataChanges(Worker_EntityId_Key EntityId, AActor* Actor)
+{
+	const bool bEnableSpatialDataDebugger = GetDefault<USpatialGDKSettings>()->IsSpatialDataDebuggerEnabled();
+
+	if (!bEnableSpatialDataDebugger || !IsServer() || !SpatialShadowActors.Contains(EntityId))
+	{
+		return;
+	}
+
+	if (SpatialShadowActors[EntityId] != nullptr)
+	{
+		SpatialShadowActors[EntityId]->CheckUnauthorisedDataChanges(EntityId, Actor);
+	}
+	else
+	{
+		UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Trying to check SpatialShadowActor which is invalid, EntityID %i"), EntityId);
+		return;
+	}
+}
+
+void USpatialNetDriver::AddSpatialShadowActor(Worker_EntityId_Key EntityId)
+{
+	const bool bEnableSpatialDataDebugger = GetDefault<USpatialGDKSettings>()->IsSpatialDataDebuggerEnabled();
+
+	if (!bEnableSpatialDataDebugger || !IsServer())
+	{
+		return;
+	}
+
+	AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(EntityId));
+	if (Actor == nullptr || !IsValid(Actor) || Actor->IsPendingKillOrUnreachable())
+	{
+		return;
+	}
+
+	if (SpatialShadowActors.Contains(EntityId))
+	{
+		UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Should only be adding a SpatialShadowActor once for each entity, EntityID %i"),
+			   EntityId);
+	}
+	else // if (!HasServerAuthority(EntityId))
+	{
+		USpatialShadowActor* SpatialShadowActor(NewObject<USpatialShadowActor>());
+		SpatialShadowActor->Init(EntityId, Actor);
+		SpatialShadowActors.Add(EntityId, SpatialShadowActor);
+	}
+}
+
+void USpatialNetDriver::RemoveSpatialShadowActor(Worker_EntityId_Key EntityId)
+{
+	const bool bEnableSpatialDataDebugger = GetDefault<USpatialGDKSettings>()->IsSpatialDataDebuggerEnabled();
+
+	if (!bEnableSpatialDataDebugger || !IsServer() || !SpatialShadowActors.Contains(EntityId))
```

Comment:
api question - would you want to warn on the last one?

Reply:
if its not a concern for this API, then checking for contains isn't necessary before calling Remove (below)